centreon/src/Core/Common/Infrastructure/Repository/DataStorageObserver.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+/*
+ * Copyright 2005 - 2023 Centreon (https://www.centreon.com/)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+
+declare(strict_types = 1);
+
+namespace Core\Common\Infrastructure\Repository;
+
+use Centreon\Domain\Repository\Interfaces\DataStorageEngineInterface;
+
+class DataStorageObserver implements DataStorageEngineInterface
+{
+    /** @var list<DataStorageEngineInterface> */
+    private array $engines = [];
+
+    /**
+     * @param DataStorageEngineInterface[] $dataStorageEngines
+     */
+    public function __construct(iterable $dataStorageEngines)
+    {
+        $this->engines = $dataStorageEngines instanceof \Traversable
+            ? iterator_to_array($dataStorageEngines)
```

Comment:
For "list", not arrays, I suggest being safer using `preserve_keys: false`/
> Note that in php 8.2+ (later for Centreon), we don't need anymore to make a difference between `Traversable` and `array` to call [`iterator_to_array`](https://www.php.net/manual/en/function.iterator-to-array.php)
```suggestion
            ? iterator_to_array($dataStorageEngines, preserve_keys: false)
```